### PR TITLE
[SPARK-11990][SQL] Don't collapse projections if UDFs are referred in multiple aliases

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-11990

We should not collapse projections in UDFs are referred in multiple aliases that will cause recomputing these UDFs.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala [hunk 1]
```diff
@@ -233,6 +236,25 @@ class UDFSuite extends QueryTest with SharedSQLContext {
           """.stripMargin).toDF(), complexData.select("m", "a", "b"))
   }
 
+  test("don't collapse UDF in order to avoid recomputing it") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: can we add an optimizer unit test instead?

Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: I have tried to do that. But as we need to use UDF, we can't do related import in the unit test such as `ProjectCollapsingSuite`. Any suggestion?

**Comment by hvanhovell** [comment]
I wonder if this is an actuall improvement. Doing back to back projections also incur a codegen and a runtime cost; which can easily be higher than the cost of (typically) two UDF calls. Have you done benchmarking on this?

This could also be taken care of by Common Subexpression Elimination.

**Comment by viirya** [comment]
@hvanhovell thanks for comment.

It might depend on the real use case. But as I can tell, ScalaUDF will involve transforming internal row to external representation during function calling. It should be more expensive.

Not sure about back to back projection and the cost it incurs. However, as the top projection simply retrieves columns from lower projection, I think it is more cheaper than UDF calls?

File: sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/UDFSuite.scala]: how about creating `ScalaUDF` manually?

**Comment by cloud-fan** [comment]
I think it's hard to tell which one is more expensive, an extra project or an extra udf evaluation? Maybe we shoud let user to decide it, use `udf((x: Int) => x + 1). nondeterministic`, which is introduced in https://github.com/apache/spark/pull/9393

cc @marmbrus @yhuai

**Comment by viirya** [comment]
@cloud-fan Hmm, I think that is reverted already?

**Comment by viirya** [comment]
Besides, from semantics, I think it is confusing because it is not actually nondeterministic.

**Comment by marmbrus** [comment]
I agree with @hvanhovell, this is already taken care of by common subexpression elimination in codegen (i.e. even though the udf appears twice we reuse the answer when doing evaluation).  As a result I think this patch is actually making things slower by adding another layer to the plan.

**Comment by viirya** [comment]
@marmbrus.Agreed. Thanks. I've not noticed there is introduced common subexpression elimination in codegen recently. Close this now.

**Pull request closed without merging**